File: Bit Manipulation/0401/0401.py
# 输入int，输出表示的二进制1的个数
def cal_bits(inter):
    c = 0
    while inter:
        if inter % 2 == 1:
            c += 1
        inter //= 2
    return c


# 每减一次1和按位与操作，就表示最低位的一个1被去除
def cal_bits2(inter):
    c = 0
    while inter:
        inter &= inter - 1
        c += 1
    return c


def func(data):
    result= []
    for h in range(12):
        for m in range(24):
            if (cal_bits(h) + cal_bits(m)) == data:
                result.append(f'{h}'+":"+f'{m}'.rjust(2,"0"))
                pass
    return result


data = 2
print(func(data))

File: Bit Manipulation/0401/0401.cpp
#include <numeric>
#include <iostream>
#include <vector>
#include <string>

int cal_bits(int inter)
{
    int c = 0;
    while (inter > 0)
    {
        inter &= inter - 1;
        c += 1;
    }
    return c;
}

std::vector<std::string> func(int data)
{
    std::vector<std::string> result;
    for (int h = 0; h < 12; h++)
        for (int m = 0; m < 60; m++)
            if (cal_bits(h) + cal_bits(m) == data)
            {
                std::string hour = std::to_string(h);
                std::string minute = m < 10 ? "0" + std::to_string(m) : std::to_string(m);
                result.emplace_back(hour+":"+minute);
            }
    return result;
}

int main(int argc, char **argv)
{
    int data = 1;
    std::vector<std::string> result = func(data);
    for(const auto &c:result)
        std::cout << c << std::endl;
    return 0;
}